pulp_python/app/tasks/publish.py

```diff
@@ -149,4 +131,27 @@ def find_artifact():
             publication=publication,
             file=File(open(metadata_relative_path, 'rb'))
         )
-        project_metadata.save()
+        project_metadata.save()  # change to bulk create when multi-table supported
+
+    ind = 0
+    current_name = index_names[ind][0]
+    package_releases = []
+    published_artifacts = []  # This list can grow up to 4 million elements for full PyPI, can be cleared if we do multiple bulk_creates
+    for release in releases.iterator():
+        if release['name'] != current_name:
+            write_project_page()
+            package_releases = []
+            ind += 1
+            current_name = index_names[ind][0]
+        content_artifact = release['contentartifact']
+        relative_path = release['filename']
+        path = f"../..{release['filename']}"
+        checksum = release['_artifacts__sha256'] or checksums[content_artifact]
+        package_releases.append((relative_path, path, checksum))
+        published_artifacts.append(models.PublishedArtifact(
+            relative_path=relative_path,
+            publication=publication,
+            content_artifact_id=content_artifact
+        ))
+    write_project_page()  # Write the final project's page
+    models.PublishedArtifact.objects.bulk_create(published_artifacts)  # maybe play around with batch size?
```

Comment:
Not so sure if this bulk_create will do well with a list of 4 million elements, but it worked fine for my test with 375,000. It's possible there is some hard-coded batch-size limit under the hood that will choose the best number for us, but I'm not sure.

Reply:
Django does do internal batching and enforces a max batch size, but I would set one manually here.  2000 would be fine.

Although - and I might be misunderstanding `pass_through`, I thought that the purpose was to avoid manually saving published artifacts?

https://github.com/django/django/blob/main/django/db/models/query.py#L547-L548

Reply:
This is from before I was setting `pass_through`. I got rid of this and the extra published_artifacts as they are no longer needed.